gps/gps.go

```diff
@@ -0,0 +1,103 @@
+// Package gps provides a driver for GPS receivers over UART and I2C
+package gps
+
+import (
+	"machine"
+	"strings"
+	"time"
+)
+
+// Device wraps a connection to a GPS device.
+type GPSDevice struct {
+	buffer   []byte
+	bufIdx   int
+	sentence strings.Builder
+	uart     machine.UART
+	bus      machine.I2C
+	address  uint16
+}
+
+// New creates a new UART GPS connection. The UART must already be configured.
+func New(uart machine.UART) GPSDevice {
+	return GPSDevice{
+		uart:     uart,
+		buffer:   make([]byte, bufferSize),
+		bufIdx:   bufferSize,
+		sentence: strings.Builder{},
+	}
+}
+
+// NewI2C creates a new I2C GPS connection.
+func NewI2C(bus machine.I2C) GPSDevice {
+	return GPSDevice{
+		bus:      bus,
+		address:  I2C_ADDRESS,
+		buffer:   make([]byte, bufferSize),
+		bufIdx:   bufferSize,
+		sentence: strings.Builder{},
+	}
+}
+
+// ReadNextSentence returns the next NMEA sentence from the GPS device.
+func (gps *GPSDevice) ReadNextSentence() (sentence string) {
+	gps.sentence.Reset()
+	var b byte = ' '
+
+	for b != '$' {
+		b = gps.readNextByte()
+	}
+
+	for b != '*' {
+		gps.sentence.WriteByte(b)
+		b = gps.readNextByte()
+	}
+	gps.sentence.WriteByte(b)
+	gps.sentence.WriteByte(gps.readNextByte())
+	gps.sentence.WriteByte(gps.readNextByte())
+
+	sentence = gps.sentence.String()
+	return sentence
+}
+
+func (gps *GPSDevice) readNextByte() (b byte) {
+	gps.bufIdx += 1
+	if gps.bufIdx >= bufferSize {
+		gps.fillBuffer()
+	}
+	return gps.buffer[gps.bufIdx]
+}
+
+func (gps *GPSDevice) fillBuffer() {
+	if (machine.I2C{}) == gps.bus {
```

Comment:
Seems like you could just make the much simpler test `if gps.uart != nil` here?

Reply:
I couldn't get that to work - it gives a compile error: ```cannot convert nil (untyped nil value) to machine.UART```

Reply:
You should probably be doing this anyhow:

```
type GPSDevice struct {
	buffer   []byte
	bufIdx   int
	sentence strings.Builder
	uart     *machine.UART
	bus      *machine.I2C
	address  uint16
}
```

Reply:
If you really want to make this nice, do like the `Buser` interface in https://github.com/tinygo-org/drivers/blob/master/ssd1306/ssd1306.go does.

Reply:
Thanks, thats what I was looking for. Will do that.

Reply:
Sounds like a good solution.
Also note that the zero value of `machine.I2C` does not necessarily mean "this is not an I2C bus". Some device may have just one I2C peripheral and for those, `machine.I2C` is zero sized (and thus always equal to the zero value).